### DISCUSSION TITLE
Behaviour for namespace connect mechanism and connect_errors

### DISCUSSION THREAD

**Discussion by Totodore** [socketio/socket.io#4970] [Q&A] [answered]
I'm implementing a [socket.io server in rust](https://github.com/totodore/socketioxide) and I'm currently adding middlewares to namespaces in order for the user to have authentication.
However I don't know what is the best between these two solutions to handle user provided errors:

* Waiting the end of the user provided handler before connecting the socket/registering the socket to the namespace or sending a connect_error in case of error. The main issue with this solution is that if the user block the task by waiting for something, the socket will only be connected after, so events triggered before the end of the task could be missed.

* Registering the socket immediately after the connect request (before the user provided handler spawn) and eventually disconnect the socket if the handler return an error. This solution implies that a socket could be temporary connected and that a connect packet would be sent and immediately after a connect_error packet.

@darrachequesne do you have an opinion on this question? What solution is used on the official implementation?

**Comment by darrachequesne** [comment, marked as answer]
Hi!

In the official implementation, we first run all the middlewares and then the socket gets connected (your 1st solution):

https://github.com/socketio/socket.io/blob/bf64870957e626a73e0544716a1a41a4ba5093bb/lib/namespace.ts#L315-L358

> so events triggered before the end of the task could be missed.

Else, a client could receive some data it should not be able to, isn't it?

**Reply by Totodore** [reply]
Sure, thanks for your quick reply :)
